File: packages/lynx-websocket/android/com/sigx/websocket/WebSocketPublisher.kt
package com.sigx.websocket

import android.util.Log
import com.lynx.react.bridge.JavaOnlyArray
import com.lynx.tasm.LynxView
import java.util.UUID

/**
 * Per-[LynxView] publisher that pumps [WebSocketEventBus] payloads into JS
 * via `LynxView.sendGlobalEvent("__sigxWebSocketEvent", [...])`.
 *
 * One instance per LynxView; instantiated by the generated
 * `GeneratedLifecyclePublishers.attachAll(lynxView)` and retained for the
 * LynxView's lifetime. The bus is global so opening a socket from one
 * LynxView and reading it from another (via the JS shim) works, but in
 * practice each LynxView holds its own JS heap so events are delivered to
 * the matching view only.
 */
class WebSocketPublisher(private val lynxView: LynxView) {

    private var token: UUID? = null

    fun attach() {
        token = WebSocketEventBus.addListener { payload ->
            try {
                val params = JavaOnlyArray()
                params.pushMap(payload)
                lynxView.sendGlobalEvent(EVENT_NAME, params)
            } catch (e: Throwable) {
                Log.w(TAG, "publish failed: ${e.message}")
            }
        }
    }

    fun detach() {
        token?.let { WebSocketEventBus.removeListener(it) }
        token = null
    }

    private companion object {
        const val TAG = "SigxWebSocketPublisher"
        const val EVENT_NAME = "__sigxWebSocketEvent"
    }
}

File: packages/lynx-websocket/ios/WebSocketPublisher.swift
import Foundation
import Lynx

/// Per-`LynxView` publisher that pumps `WebSocketEventBus` payloads into JS
/// via `LynxView.sendGlobalEvent("__sigxWebSocketEvent", [...])`.
///
/// One instance per LynxView; instantiated by the generated
/// `GeneratedLifecyclePublishers.attachAll(to:)` and retained for the
/// LynxView's lifetime. The bus is global so opening a socket from one
/// LynxView and reading it from another (via the JS shim) works, but in
/// practice each LynxView holds its own JS heap so events are delivered to
/// the matching view only.
final class WebSocketPublisher {

    private weak var lynxView: LynxView?
    private var token: UUID?

    init(lynxView: LynxView) {
        self.lynxView = lynxView
        self.token = WebSocketEventBus.shared.addListener { [weak self] payload in
            guard let view = self?.lynxView else { return }
            // sendGlobalEvent expects an array of params; pass a single
            // dictionary as the only param, matching what the JS shim
            // expects (it reads the first arg as `NativeEvent`).
            view.sendGlobalEvent("__sigxWebSocketEvent", withParams: [payload])
        }
    }

    deinit {
        if let token = token {
            WebSocketEventBus.shared.removeListener(token)
        }
    }
}
